contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,350 @@
+
+This type of race can only occur when CSI and in-tree components are acting in
+reaction to an outside object. This happens at two points: if downgrade is
+performed after a volumeattachment but before the volume is actually attached,
+and if PVC is created for dynamically provisioned and cluster is downgraded
+before provisioning.  In both cases the in-tree plugin and CSI Driver may race
+by both reacting to the external object without knowledge of each other.
+
+For Provisioning this is solved by having the CSI Driver have a different
+provisioner name than the in-tree provisioner. A PVC can only be marked for
+provisioning for one of them at a time.
+
+The VolumeAttachment case is a little more tricky as volumes attached by in-tree
+plugin won’t have VolumeAttachment instance after upgrade. CSI plugin will try
+to attach the volume, create VolumeAttachment and call CSI driver. It’s
+important that the CSI driver is idempotent and the external attacher uses the
+same parameters as the internal volume plugin used to ensure the idempotency.
+Especially CSI VolumeHandle must be carfully chosen! Volumes attached by CSI
+driver will have VolumeAttachment. In-tree volume plugin should be able to
+detach the volume in the usual way after downgrade, however, A/D controller
+should delete the VolumeAttachment.
+
+For mounts, there are many issues with live upgrade/downgrade due to differences
+of paths and handling of directories used for MountDevice and SetUpAt. For now
+this remains unsolved.
+
+THE CURRENT RECOMMEND SOLUTION IS A BIG CAVEAT: NODES MUST BE DRAINED ON A
+KUBELET UPGRADE/DOWNGRADE OR CSI MIGRATION IS TURNED ON/OFF
+
+## Cloud Provider Requirements
+There is a push to remove CloudProvider code from kubernetes.
+
+There will not be any general auto-deployment mechanism for ALL CSI drivers
+covered in this document so the timeline to remove CloudProvider code using this
+design is undetermined: For example: At some point GKE could auto-deploy the GCE
+PD CSI driver and have migration for that turned on by default, however it may
+not deploy any other drivers by default. And at this point we can only remove
+the code for the GCE In-tree plugin (this would still break anyone doing their
+own deployments while using GCE unless they install the GCE PD CSI Driver).
+
+We could have auto-deploy depending on what cloud provider kubernetes is running
+on. But AFAIK there is no standard mechanism to guarantee this on all Cloud
+Providers.
+
+For example the requirements for just the GCE Cloud Provider code for storage
+with minimal disruption to users would be:
+* In-tree to CSI Plugin migration goes GA
+* GCE PD CSI Driver deployed on GCE/GKE by default (resource requirements of
+  driver need to be determined)
+* GCE PD CSI Migration turned on by default
+* Remove in-tree plugin code and cloud provider code
+
+And at this point users doing their own deployment and not installing the GCE PD
+CSI driver encounter an error.
+
+## Testing
```

Comment:
I dont think we will be able to support configuration skew.
We will have this devicePath incompatibility, same issue as above (https://github.com/kubernetes/community/pull/2199/files#r192096628) except now theres no actual solution because versions are actually different.

I will add a section about disabling this feature and ignoring the feature gates if Master/Node are not both on a version that supports Migration.

Reply:
added a section about this

Reply:
I am still not sure how it safely migrate. Admin disables the feature and updates cluster. Now all components have the right version, however, they still use in-tree plugins. What now?

1. Admin enables the feature on masters. They will start re-attaching volumes and creating VolumeAttachment objects.
2. For each node:
  2.1 Admin drains the node.
  2.2 Admin enables the feature gate.
  2.3 Admin starts the node.

At point 2.1, the evicted pods are "moved" to another nodes (i.e. they're killed and ReplicaSet/StatefulSet re-creates new ones somewhere else). The new node may or may not have the feature enabled. But master is already enabled, so it will go the CSI way and create VolumeAttachment and only the migrated node can understand it.

Is this reliable?

Just brainstorming here: can there be a flag (annotation) on a node that would tell if the feature (and which plugins) are migrated to CSI on the node? So A/D controller can either go the CSI way or internal plugin *per node*? It would be really complicated on the A/D controller side. We could make it easier if we allow to flip the feature only on drained nodes so A/D controller does not need to re-attach all volumes and create VolumeAttachments, because there is nothing to re-attach.

Reply:
I think it is better here to do not enable it on A/D Controller unless the `lowest node verson is > X && migration enabled`. This would solve the issue you bring up here. It may be possible to do in on a node by node basis but I think that would be far too complex of a problem to solve for this problem that only exists in a very transitory state (1-2 release cycles)

Reply:
Nevermind, I also see an issue with this that the newly upgraded node will be `version X && migration enabled` and has no way to know (?) what versions the other nodes are on. So that node cannot make a "decision" about whether to use migration or not.

In this case the A/D controller may have migration "off" because it sees there are still nodes at version X-1. But the node at version X will turn migration "on" because of the above reasoning (unless it can tell other node versions). 

I dont really follow your sentence: "We could make it easier if we allow to flip the feature only on drained nodes so A/D controller does not need to re-attach all volumes and create VolumeAttachments, because there is nothing to re-attach."

Does this solve the problem I just outlined?

Reply:
Let me rephrase my suggestion. When kubelet starts, it checks what feature gates are enabled and annotates itself with say `csi.attach.kubernetes.io/gce-pd=true` to tell A/D controller that it supports GCE PD through CSI.

A/D controller has the feature enabled too and before it starts looking for a volume plugin, it checks node annotation. If the node says it supports CSI for GCE-PD, it will create VolumeAttachment. If the node does not support CSI, A/D controller uses the old volume plugin to attach the volume (and set node.status.volumesAttached).

So the A/D controller must instantiate both old volume plugins and their CSI counterparts and have possibility to use both at the same time (for different nodes). This should not be that complicated, IMO.

Now, what if admin flips the feature gate in *one* kubelet (e.g. because it's getting updated or the admin enables the feature on each node)? There are volumes attached to the node. Here I suggest to drain the node first, i.e. unmount and detach all volumes and flip the feature gate (and restart kubelet) only when the node is completely empty. Kubelet could spit some warnings / errors when it detects a volume attached by the old volume plugin.

Reply:
I see, I will add some steps for this in the design. I am curious would this cause a race between adding the annotation and a volume being attached. For example if A/D Controller begins attach on the node after it is restarted but before the node is annotated, it will attach with in-tree plugin and then the kubelet will be trying to mount with CSI (and probably failing).

I guess this will be reconciled by the other part we discussed doing the re-attach during the reconcile loops but it seems like a lot to reason about. I can't think of a better way.

Reply:
See new commit for added information

Reply:
I've actually reworked the entire section into `## Upgrade/Downgrade, Migrate/Un-migrate`. PTAL. I believe it succinctly covers all the bases we have discussed so far. I have included an upgrade and downgrade workflow in `Upgrade/Downgrade Migrate/Unmigrate Scenarios` to illustrate the solution